### ISSUE TITLE
[spline-reco] Bug leads to segmentation violation in case of an 'anomalously large charge' (Trac #2119)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#2111]
In `I3SplineRecoLikelihood::KS0Limit(int npulses, int level)` an array is accessed with `ks_test[npulses - 1]` but does not check if `npulses < 1`.

I saw a couple of events that call this function with `npulses = 0` but apparently it didn't break the execution. However, I found a case with

```text
I3HitOm.h:111 WARNING: got anomalously large charge=1.64923e+10 PE in OM at (x,y,z)=(429.8,351.0,-344.9)
```

that leads to an overflow of `totalcharge` in `I3SplineRecoLikelihood::GetApprovedCharge`. Finally, `npulses` is small enough (e.g. `npulses = -2147483648`) to cause a segmentation violation.

http://code.icecube.wisc.edu/projects/icecube/browser/IceCube/projects/spline-reco/trunk/private/spline-reco/I3SplineRecoLikelihood.cxx?rev=157082#L867

This bug blocks processing of a pass2 file.

Migrated from https://code.icecube.wisc.edu/ticket/2119
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:14:55", 
    "description": "In `I3SplineRecoLikelihood::KS0Limit(int npulses, int level)` an array is accessed with `ks_test[npulses - 1]` but does not check if `npulses < 1`.\n\nI saw a couple of events that call this function with `npulses = 0` but apparently it didn't break the execution. However, I found a case with\n\n{{{\nI3HitOm.h:111 WARNING: got anomalously large charge=1.64923e+10 PE in OM at (x,y,z)=(429.8,351.0,-344.9)\n}}}\n\nthat leads to an overflow of `totalcharge` in `I3SplineRecoLikelihood::GetApprovedCharge`. Finally, `npulses` is small enough (e.g. `npulses = -2147483648`) to cause a segmentation violation.\n\nhttp://code.icecube.wisc.edu/projects/icecube/browser/IceCube/projects/spline-reco/trunk/private/spline-reco/I3SplineRecoLikelihood.cxx?rev=157082#L867\n\nThis bug blocks processing of a pass2 file.", 
    "reporter": "joertlin", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1550067295757382", 
    "component": "analysis", 
    "summary": "[spline-reco] Bug leads to segmentation violation in case of an 'anomalously large charge'", 
    "priority": "critical", 
    "keywords": "", 
    "time": "2017-12-01T21:22:27", 
    "milestone": "", 
    "owner": "kjmeagher", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20171201T21:34:59: cweaver commented:

> I think this would be better?
> 
> 
> ```text
> Index: private/spline-reco/I3SplineRecoLikelihood.cxx
> ===================================================================
> --- private/spline-reco/I3SplineRecoLikelihood.cxx	(revision 159227)
> +++ private/spline-reco/I3SplineRecoLikelihood.cxx	(working copy)
> @@ -824,7 +824,7 @@
>  }
>  
>  double
> -I3SplineRecoLikelihood::KS0Limit (int npulses, int level)
> +I3SplineRecoLikelihood::KS0Limit (double npulses, int level)
>  {
>      /**
>       * first  column = level 5 corresponds to 80% confidence level
> @@ -862,9 +862,18 @@
>      static const double a[5] = { 1.07, 1.14, 1.22, 1.36, 1.63 };
>      double limit;
>  
> +    if(npulses<0)
> +    {
> +        npulses=0;
> +    }
>      if (npulses <= 20)
>      {
> -        limit = ks_test[npulses - 1][5 - level];
> +        int index=(int) (npulses + 0.5) - 1;
> +        if(index < 0)
> +        {
> +            index=0;
> +        }
> +        limit = ks_test[index][5 - level];
>      }
>      else
>      {
> @@ -935,7 +944,7 @@
>                  diff = fabs (Cdf - runningcharge / totalcharge);
>              kolmogorov_maximum = std::max (diff, kolmogorov_maximum);
>          }
> -        while (kolmogorov_maximum > KS0Limit ((int) (totalcharge + 0.5), confidence_level) && --LastPulse != idom->second.begin ())
> +        while (kolmogorov_maximum > KS0Limit (totalcharge, confidence_level) && --LastPulse != idom->second.begin ())
>          {			//if confidence level is not met subtract the last pulse until it is but always keep the first pulse
>              totalcharge -= LastPulse->GetCharge ();
>              kolmogorov_maximum = 0.;
> Index: private/spline-reco/I3SplineRecoLikelihood.h
> ===================================================================
> --- private/spline-reco/I3SplineRecoLikelihood.h	(revision 159227)
> +++ private/spline-reco/I3SplineRecoLikelihood.h	(working copy)
> @@ -141,7 +141,7 @@
>          void GetNoiseWeights (double &physicsWeight, double &randomWeight, double &afterWeight, double &preLateWeight, double expected_pes);
>  
>          /// Kolmogorov-Smirnov limits
> -        double KS0Limit (int npulses, int level);
> +        double KS0Limit (double npulses, int level);
>  
>          /// returns only the charge of a dom that fits to pdf by performing Kolmogorov-Smirnov-Test
>          double GetApprovedCharge (I3RecoPulseSeriesMap::const_iterator idom, const I3Particle p, const double geotime, const double mean_pes, const double stochMean_pes);
> ```
> 
> This keeps large charges as a double to pass to `sqrt`, and also avoids the underflow issue for small charges.

**Comment by icecube-trac** [comment]
Trac update at 20171205T07:44:14: 
* kjmeagher changed status from "new" to "accepted"
* kjmeagher changed owner from "olivas" to "kjmeagher"
* kjmeagher changed milestone from "December 2017" to "IceRec 5.2"

**Comment by icecube-trac** [comment]
Trac update at 20171205T07:44:26: 
* kjmeagher changed resolution from "" to "fixed"
* kjmeagher changed status from "accepted" to "closed"
* kjmeagher commented:

> In changeset 159914:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="159914"
> handle Npulses as double to prevent overflow, fixes #2111
> ```

**Comment by icecube-trac** [comment]
Trac update at 20171208T16:16:50: kjmeagher commented:

> In changeset 159995:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="159995"
> handle Npulses as double to prevent overflow, fixes #2111
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:14:55: 
* olivas commented:

> Milestone IceRec 5.2 deleted


* olivas changed milestone from "IceRec 5.2" to ""
